e2e/src/test/java/no/sysco/testing/kafka/e2e/ContainersIT.java

```diff
@@ -0,0 +1,194 @@
+            .waitingFor(Wait.forHttp("/subjects").forStatusCode(200));
+    schemaRegistry.start();
+    SCHEMA_REGISTRY_INSIDE_DOCKER_ENV =
+        "http://" + schemaRegistry.getNetworkAliases().get(0) + ":8081";
+  }
+
+  private static void setJsonServer() {
+    jsonServer =
+        new GenericContainer("zhenik/json-server")
+            .withExposedPorts(80)
+            // all containers put in same network
+            .withNetwork(kafka.getNetwork())
+            .withEnv("ID_MAP", "id")
+            .withNetworkAliases("json-server")
+            .withClasspathResourceMapping(
+                "json-server-database.json", "/data/db.json", BindMode.READ_WRITE)
+            .waitingFor(Wait.forHttp("/").forStatusCode(200));
+
+    jsonServer.start();
+    // provide availability make http calls from localhost against docker env
+    JSON_SERVER_EXPOSED = LOCAL_HOST + ":" + jsonServer.getMappedPort(80);
+    JSON_SERVER_INSIDE_DOCKER_ENV = "http://" + jsonServer.getNetworkAliases().get(0) + ":80";
+  }
+
+  private static void setHttProducer() {
+    httpProducer =
+        new GenericContainer("zhenik/http-producer:data-pipeline")
+            .withExposedPorts(8080)
+            .withNetwork(kafka.getNetwork())
+            .withEnv("APPLICATION_PORT", "8080")
+            .withEnv("KAFKA_BOOTSTRAP_SERVERS", KAFKA_BROKER_INSIDE_DOCKER_ENV)
+            .withEnv("SCHEMA_REGISTRY_URL", SCHEMA_REGISTRY_INSIDE_DOCKER_ENV)
+            .withEnv("SINK_TOPIC", TOPIC)
+            .waitingFor(Wait.forHttp("/messages").forStatusCode(200));
+    httpProducer.start();
+
+    HTTP_PRODUCER_EXPOSED = LOCAL_HOST + ":" + httpProducer.getMappedPort(8080);
+  }
+
+  private static void setHttpMaterializer() {
+
+    createTopic(TOPIC);
+
+    String messageEndpoint = "http://json-server/messages";
+    System.out.println("HERE: " + messageEndpoint);
+    httpMaterializer =
+        new GenericContainer("zhenik/http-materializer:data-pipeline")
+            .withNetwork(kafka.getNetwork())
+            .withEnv("KAFKA_BOOTSTRAP_SERVERS", KAFKA_BROKER_INSIDE_DOCKER_ENV)
+            .withEnv("SCHEMA_REGISTRY_URL", SCHEMA_REGISTRY_INSIDE_DOCKER_ENV)
+            .withEnv("DATABASE_REST_SERVICE_URL", messageEndpoint)
+            .withEnv("SOURCE_TOPIC", TOPIC);
+    httpMaterializer.start();
+  }
+
+  private static void createTopic(String topicName) {
+    // kafka container uses with embedded zookeeper
+    // confluent platform and Kafka compatibility 5.1.x <-> kafka 2.1.x
+    // kafka 2.1.x require option --zookeeper, later versions use --bootstrap-servers instead
+    String createTopic =
+        String.format(
+            "/usr/bin/kafka-topics --create --zookeeper localhost:2181 --replication-factor 1 --partitions 1 --topic %s",
+            topicName);
+    try {
+      final Container.ExecResult execResult = kafka.execInContainer("/bin/sh", "-c", createTopic);
+      if (execResult.getExitCode() != 0) fail();
+    } catch (Exception e) {
```

Comment:
I this case, I expect **NO** exception, if any occur or cmd result execution different than `0` ->  I force test to `fail()`

Reply:
why not using adminclient api? wouldn't that be a bit cleaner?

Reply:
I want to show different approaches of how test containers can be used. Otherwise, to use admin client api I need  to include topic creation in code-base